frame/benchmarking/src/lib.rs

```diff
@@ -903,6 +928,85 @@ macro_rules! impl_benchmark_test {
 	};
 }
 
+/// This creates a test suite which runs the module's benchmarks.
+///
+/// When called in [`pallet_example`] as
+///
+/// ```rust,ignore
+/// impl_benchmark_test_suite!(crate::tests::new_test_ext, crate::tests::Test);
+/// ```
+///
+/// It expands to the equivalent of:
+///
+/// ```rust,ignore
+/// #[cfg(test)]
+/// mod tests {
+/// 	use super::*;
+/// 	use crate::tests::{new_test_ext, Test};
+/// 	use frame_support::assert_ok;
+///
+/// 	#[test]
+/// 	fn test_benchmarks() {
+/// 		new_test_ext().execute_with(|| {
+/// 			assert_ok!(test_benchmark_accumulate_dummy::<Test>());
+/// 			assert_ok!(test_benchmark_set_dummy::<Test>());
+/// 			assert_ok!(test_benchmark_another_set_dummy::<Test>());
+/// 			assert_ok!(test_benchmark_sort_vector::<Test>());
+/// 		});
+/// 	}
+/// }
+/// ```
+///
+/// ## Arguments
+///
+/// The first argument, `new_test_ext`, must be the path to a function which takes no arguments
+/// and returns either a `sp_io::TestExternalities`, or some other type with an identical interface.
+///
+/// The second argument, `test`, must be the path to the runtime. The item to which this must refer
+/// will generally take the form:
+///
+/// ```rust,ignore
+/// frame_support::construct_runtime!(
+/// 	pub enum Test where ...
+/// 	{ ... }
+/// );
+/// ```
+///
+// ## Notes (not for rustdoc)
+//
+// The biggest challenge for this macro is communicating the actual test functions to be run. We
+// can't just build an array of function pointers to each test function and iterate over it, because
+// the test functions are parameterized by the `Test` type. That's incompatible with
+// monomorphization: if it were legal, then even if the compiler detected and monomorphized the
+// functions into only the types of the callers, which implementation would the function pointer
+// point to? There would need to be some kind of syntax for selecting the destination of the pointer
+// according to a generic argument, and in general it would be a huge mess and not worth it.
+//
+// Instead, we're going to steal a trick from `fn run_benchmark`: generate a function which is
+// itself parametrized by `Test`, which accepts a `&[u8]` parameter containing the name of the
+// benchmark, and dispatches based on that to the appropriate real test implementation. Then, we can
+// just iterate over the `Benchmarking::benchmarks` list to run the actual implementations.
+#[macro_export]
+macro_rules! impl_benchmark_test_suite {
+	($new_test_ext:path, $test:path) => {
+		#[cfg(test)]
+		mod tests {
+			use super::*;
+			use $crate::frame_support::assert_ok;
+
+			#[test]
+			fn test_benchmarks() {
+				$new_test_ext().execute_with(|| {
+					use $crate::Benchmarking;
+					for benchmark_name in Module::<$test>::benchmarks(true) {
+						assert_ok!(test_bench_by_name::<$test>(benchmark_name));
```

Comment:
I am concerned with this approach if one of the benchmarks fails, it will be hard to identify which benchmark is the issue.

I would prefer we actually generate N tests in this macro where each benchmark gets its own entire function, and thus when it fails, it is easy to identify which test failed.

Reply:
This is critical for debugging during benchmark development.

Reply:
Like here: https://github.com/paritytech/substrate/blob/master/frame/assets/src/benchmarking.rs#L237

Reply:
> I would prefer we actually generate N tests in this macro where each benchmark gets its own entire function

I'd prefer that too. Unfortunately, it's hard: we don't have a simple way to iterate over the names of the benchmarks within the macro scope which generates the test module, which is why we introduce the `test_bench_by_name` function at all. 

I'll think some more about how to accomplish this.

Reply:
moving the assert to inside test_bench_by_name should be equivalent to what was before no ?

Reply:
No: a failing assertion within `test_bench_by_name` still fails the single test without testing anything else, or propagating information about which test failed.

A short-term hack for identifying failing benchmarks would be to simply print the current benchmark to stdout: you can identify the failing benchmark because its name will be the last printed. On success, `cargo test` will just swallow stdout.

However, doing this properly involves building the functionality of `impl_benchmark_test_suite` directly into the `benchmarks` macro; there just isn't another way that I can see to generate individual test cases without requiring the user to explicitly enumerate the generated test functions. Obviously, explicitly enumerating those is worse than the status quo: it retains the boilerplate, but makes it less clear what is being accomplished.

If I have to substantially modify the `benchmarks` macro, I'd like to rewrite it as a procedural macro, because that's just much easier to understand than some 900 lines of interlinked recursive example macros as currently exist. However, that feels like kind of an enormous job. I can say that it'll take me a fair while to complete it.

Reply:
Before I start on a huge project rewriting the `benchmarks` macro, what do you think of cb6a33d4f? It fails at the goal of generating a single test case per benchmark test, but it succeeds at the goals of running all the tests, and providing a list of failing tests, and what the failures were.

Reply:
it is hard for me to tell without a UI test, but I can be fine with it

Reply:
hmm actually a lot of code in benchmark would simply do assert and unwrap instead of returning error.
For them maybe we can print the name of the test before its exeuction

Reply:
Good point. fd713121b